Add Bias and Boundary Layer Features to PCCMG and PRMG

miaoyinb opened the pull request:
closes #20637

<!--
If this PR is associated with an issue be sure to reference it
by including "refs #<issue>" or "closes #<issue>" (e.g., #closes #1234).

If this PR implements an enhancement that does not have an existing issue, please add the following:
-->

## Reason
<!--Why do you need this feature or what is the enhancement?-->

## Design
<!--A concise description (design) of the enhancement.-->

## Impact
<!--Will the enhancement change existing APIs or add something new?-->


<!--
If this PR implements a bug fix, please create an issue for the bug and reference the issue.
-->

miaoyinb committed:
Add Bias and Boundary Layer Features to PCCMG and PRMG

closes #20637

moosebuild commented on large_media:
```diff
@@ -1 +1 @@
-Subproject commit acc498685d1453fda11c2000886fdacdae6ae333
+Subproject commit dd3bc61c9f9e9d235d2810e8efe924a78ca97f75
```
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/994324/) on a817762 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/20638/site)

This comment will be updated on new commits.

miaoyinb commented:
@eshemon @shikhar413 @aprilnovak  FYI

@loganharbour @GiudGiud Add radial bias option for the two Reactor module MGs. Thank you in advance for your review.

GiudGiud commented on modules/reactor/include/meshgenerators/PolygonMeshGeneratorBase.h:
```diff
@@ -401,4 +468,59 @@ class PolygonMeshGeneratorBase : public MeshGenerator
+   * Creates bias terms for multiple blocks.
+   * @param radial_biases bias growth factors of the main regions of the blocks
+   * @param intervals radial interval numbers of the main regions of the blocks
+   * @param inner_boundary_layer_fractions radial fractions of the blocks to be defined as the
+   * boundary layers near the inner boundaries
+   * @param inner_boundary_layer_intervals radial interval numbers of the blocks to be defined as
+   * the boundary layers near the inner boundaries
+   * @param inner_boundary_layer_biases bias growth factors of the blocks to be defined as the
+   * boundary layers near the inner boundaries
+   * @param outer_boundary_layer_fractions radial fractions of the blocks to be defined as the
+   * boundary layers near the outer boundaries
+   * @param outer_boundary_layer_intervals radial interval numbers of the blocks to be defined as
+   * the boundary layers near the outer boundaries
+   * @param outer_boundary_layer_biases bias growth factors of the blocks to be defined as the
+   * boundary layers near the outer boundaries
```
you say `of the blocks` but I think you mean elements?

miaoyinb replied:
You are right. Should be `of the elements within the blocks`

GiudGiud reviewed (commented):
I ll review more later

miaoyinb committed:
Minor documentation revision 1 ref #20637

Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

miaoyinb committed:
Minor doxygen comments revision 1 ref #20637

moosebuild commented on large_media:
```diff
@@ -1 +1 @@
-Subproject commit acc498685d1453fda11c2000886fdacdae6ae333
+Subproject commit 5e3373c5383afdaa2dafca6495923348131f0b22
```
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

miaoyinb committed:
Remove redundant lines ref #20637

GiudGiud reviewed (changes requested):
one more file to go

miaoyinb committed:
use struct for boundary layer parameters ref #20637

miaoyinb committed:
Add comment for peripheral ring boundary layer interpolation ref #20637

miaoyinb committed:
Use std::make_unique instead of libmesh_make_unique ref #20637

miaoyinb commented on modules/reactor/src/meshgenerators/PeripheralRingMeshGenerator.C:
```diff
@@ -162,31 +230,171 @@ PeripheralRingMeshGenerator::generate()
+    }
+    // Establish interpolation
+    linterp_x = libmesh_make_unique<LinearInterpolation>(interp_azi_data, interp_x_data);
+    linterp_y = libmesh_make_unique<LinearInterpolation>(interp_azi_data, interp_y_data);
```
X and Y need to be interpolated. I want to keep the azimuthal angles unchanged for the inner boundary layer (otherwise the mesh either look weird or messed). Thus, I use normal surface direction to define the location of the outer boundary of the boundary layer (by defining key Point positions). These key positions do not have the original azimuthal angles, so they are just used as interpolation input.

miaoyinb committed:
Add angle unit to the docstring of radiusCorrectionFactor() ref #20637

Error message typo fix ref #20637

Docstring improvements and typo fixes ref #20637

nest tests and typo fixes ref #20637

clang-format fix ref #20637

miaoyinb commented on modules/reactor/src/meshgenerators/PolygonConcentricCircleMeshGeneratorBase.C:
```diff
@@ -250,6 +468,76 @@ PolygonConcentricCircleMeshGeneratorBase::PolygonConcentricCircleMeshGeneratorBa
+                   "Summation of duct_inner_boundary_layer_fractions and "
+                   "duct_outer_boundary_layer_fractions cannot exceeds the duct layer width.");
+    }
+  }
```
It is checked, near the end of constructor

miaoyinb commented on modules/reactor/src/meshgenerators/PeripheralRingMeshGenerator.C:
```diff
@@ -162,31 +230,171 @@ PeripheralRingMeshGenerator::generate()
+            (points_array[_peripheral_inner_boundary_layer_intervals + _peripheral_layer_num][i] -
+             origin_pt)
+                .norm()))
+      paramError("peripheral_inner_boundary_layer_width",
```
This check needs to be done after collecting all the external boundary nodes info and polygonization correction. So it has to be here.

GiudGiud commented on modules/reactor/src/meshgenerators/PolygonMeshGeneratorBase.C:
```diff
@@ -34,25 +35,98 @@ PolygonMeshGeneratorBase::generate()
+      background_inner_boundary_layer_bias, background_inner_boundary_layer_intervals);
+  const auto outer_background_bias_terms = biasTermsCalculator(
+      background_outer_boundary_layer_bias, background_outer_boundary_layer_intervals);
+  auto rings_bias_terms = biasTermsCalculator(ring_radial_biases,
+                                              rings,
+                                              ring_inner_boundary_layer_fractions,
+                                              ring_inner_boundary_layer_intervals,
+                                              ring_inner_boundary_layer_biases,
+                                              ring_outer_boundary_layer_fractions,
+                                              ring_outer_boundary_layer_intervals,
+                                              ring_outer_boundary_layer_biases);
```
this makes me think the struct could even hold 6 members? Not sure it's true though, background has less params?

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/999410/) on 59e7064 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/20638/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/20638/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format c90b1f04c38b5fa8c35d8b16e7ca4ae7c7ad7115`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/999410/) on 59e7064 wanted to post the following:

The following file(s) are changed:

  conda/mpich/conda_build_config.yaml

The conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/999410/) on 59e7064 wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/999441/) on eab3c6c wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

miaoyinb committed:
More comment to explan use of inverse bias ref #20637

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/999472/) on 2195dee wanted to post the following:

The following file(s) are changed:

  conda/mpich/conda_build_config.yaml

The conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/999472/) on 2195dee wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

GiudGiud commented on modules/reactor/src/meshgenerators/PolygonMeshGeneratorBase.C:
```diff
@@ -34,25 +35,98 @@ PolygonMeshGeneratorBase::generate()
+                                             duct_outer_boundary_layer_intervals,
+                                             duct_outer_boundary_layer_biases);
+
+  std::vector<unsigned int> total_rings;
```
the equivalent param for ducts is duct_layers right?

miaoyinb replied:
Yes. May be use `ring_layers`?

GiudGiud replied:
seems good to me

GiudGiud commented on modules/reactor/src/meshgenerators/PolygonMeshGeneratorBase.C:
```diff
@@ -34,25 +35,89 @@ PolygonMeshGeneratorBase::generate()
+    total_rings.push_back(rings[i] + ring_inner_boundary_layer_params.intervals[i] +
+                          ring_outer_boundary_layer_params.intervals[i]);
+
+  if (background_inner_boundary_layer_params.intervals)
```
so you're tagging the inner background boundary layer onto the pin rings
and the outer background boundary layer onto the duct rings. 

Why is that?

miaoyinb replied:
After discussing with TH people, it seems that boundary layer needs to have a uniform thickness for better TH code performance and analyses. Namely, they need to be conformal to the boundary shape. In this case, the inner boundary layer needs to be circular and the outer boundary layer needs to be "duct" shape. So the easiest approach is to mesh them as if they are rings/ducts.

GiudGiud replied:
thanks for the explanation

GiudGiud commented on modules/reactor/src/meshgenerators/PolygonMeshGeneratorBase.C:
```diff
@@ -338,6 +419,7 @@ void
 PolygonMeshGeneratorBase::backgroundNodes(ReplicatedMesh & mesh,
                                           const unsigned int num_sectors_per_side,
                                           const unsigned int background_intervals,
+                                          const std::vector<Real> biased_terms,
```
biased terms are no longer 1 for no-bias right?
They are 1/number of intervals right?

miaoyinb replied:
Yes.

GiudGiud commented on modules/reactor/include/meshgenerators/PolygonMeshGeneratorBase.h:
```diff
@@ -401,4 +448,59 @@ class PolygonMeshGeneratorBase : public MeshGenerator
+   * the boundary layer near the outer boundary
+   * @param outer_boundary_layer_bias bias growth factor of the elements within the block to be
+   * defined as the boundary layer near the outer boundary
+   * @return bias terms describing radial node positions of a single block
```
let's be more precise here. 
This is the fraction of the total radius for each node right?
This is cumulative too right? As in b[n] = b[n-1] + radial distance between node n and n-1

miaoyinb replied:
Yes. will modify this.

GiudGiud reviewed (changes requested):
It seems there are more image files than needed in the large_media submodule. Is this for another PR? 

Good to go as far as I can tell once this is addressed. 
I m glad this PR came separately, this does make the objects more complicated, but for a good reason. I'm not sure how to do it more clearly though. Maybe if there's opportunities to move some code from PeripheralRingMG::generate() into separate routines with a well-defined purpose?

GiudGiud commented:
Dont worry about the mac end user failure btw, it's a WIP to fix it

miaoyinb commented:
> It seems there are more image files than needed in the large_media submodule. Is this for another PR?
> 
> Good to go as far as I can tell once this is addressed. I m glad this PR came separately, this does make the objects more complicated, but for a good reason. I'm not sure how to do it more clearly though. Maybe if there's opportunities to move some code from PeripheralRingMG::generate() into separate routines with a well-defined purpose?

The `large_media` also has images for the other PR I opened. Let me take a look to see if anything can be moved out of generate().

miaoyinb committed:
precise doxygen comments for BiasTermCalculator ref #20637

miaoyinb committed:
Better names for rings and total_rings ref #20637

miaoyinb committed:
Move inner boundary layer nodes position definer to a routine ref #20637

Apply suggestions from code review

test details improvements ref #20637

Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

Apply suggestions from code review

Another test details improvements ref #20637

Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

typo fix ref #20637

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/999820/) on 8475665 wanted to post the following:

The following file(s) are changed:

  conda/mpich/conda_build_config.yaml

The conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/999820/) on 8475665 wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

miaoyinb commented:
@GiudGiud `Internal app tests` failed. This is not due to my commits right?

GiudGiud commented:
No it s not it s my siblings transfer PR that broke everything. Should be resolved now though

miaoyinb commented:
Let me know if any additional revision is needed for this PR. Thanks. @GiudGiud

GiudGiud commented on modules/reactor/test/tests/meshgenerators/polygon_concentric_circle_mesh_generator/tests:
```diff
@@ -85,171 +126,319 @@
+      cli_args = 'Mesh/gen/background_block_ids="20"
+                  --mesh-only "poly_2d_err.e"'
+      expect_err = 'This parameter must be either unset or have a length of two when ring_radii is not provided and background_intervals is not unity.'
+      detail = "background_block_ids has a length other than zero or two when the mesh does not have rings and background_intervals equals one."
```
```suggestion
      detail = "background_block_ids has a length other than zero or two when the mesh does not have rings and background_intervals does not equal one."
```

GiudGiud commented on modules/reactor/test/tests/meshgenerators/polygon_concentric_circle_mesh_generator/tests:
```diff
@@ -85,171 +126,319 @@
+      cli_args = 'Mesh/gen/background_block_names="back1"
+                  --mesh-only "poly_2d_err.e"'
+      expect_err = 'This parameter must be either unset or have a length of two when ring_radii is not provided and background_intervals is not unity.'
+      detail = "background_block_names has a length other than zero or two when the mesh does not have rings and background_intervals equals one."
```
```suggestion
      detail = "background_block_names has a length other than zero or two when the mesh does not have rings and background_intervals does not equal one."
```

GiudGiud commented on modules/reactor/test/tests/meshgenerators/polygon_concentric_circle_mesh_generator/tests:
```diff
@@ -85,171 +126,319 @@
+    [err_odd_num_sectors_per_side]
+      type = 'RunException'
+      input = 'poly_2d.i'
+      cli_args = 'Mesh/gen/num_sectors_per_side="3 3 3 3 3"
```
```suggestion
      cli_args = 'Mesh/gen/num_sectors_per_side="2 2 2 2 3"
```
no need to overdo it

GiudGiud commented on modules/reactor/test/tests/meshgenerators/polygon_concentric_circle_mesh_generator/tests:
```diff
@@ -85,171 +126,319 @@
+                  Mesh/gen/duct_block_ids="20"
+                  --mesh-only "poly_2d_err.e"'
+      expect_err = 'This parameter must have the same length as duct_intervals if set.'
+      detail = "duct_block_ids size is not consistent with number of ducts in the mesh."
```
```suggestion
      detail = "duct_block_ids size is not consistent with the number of ducts in the mesh."
```

GiudGiud commented on modules/reactor/test/tests/meshgenerators/polygon_concentric_circle_mesh_generator/tests:
```diff
@@ -85,171 +126,319 @@
+                  Mesh/gen/duct_block_names="duct1"
+                  --mesh-only "poly_2d_err.e"'
+      expect_err = 'This parameter must have the same length as duct_intervals if set.'
+      detail = "duct_block_names size is not consistent with number of ducts in the mesh."
```
```suggestion
      detail = "duct_block_names size is not consistent with the number of ducts in the mesh."
```

GiudGiud commented on modules/reactor/test/tests/meshgenerators/polygon_concentric_circle_mesh_generator/tests:
```diff
@@ -85,171 +126,319 @@
+      cli_args = 'Mesh/gen/background_inner_boundary_layer_width=5.0
+                  Mesh/gen/background_outer_boundary_layer_width=5.0
+                  --mesh-only "poly_2d_err.e"'
+      expect_err = 'must be less then the minimum thickness of the background region.'
```
```suggestion
      expect_err = 'must be less than the minimum thickness of the background region.'
```
please fix the error message too. Sorry to be so nitpicky, I'm doing this as I m looking for other potential more important issues

GiudGiud commented on modules/reactor/test/tests/meshgenerators/polygon_concentric_circle_mesh_generator/tests:
```diff
@@ -85,171 +126,319 @@
+      input = 'poly_2d.i'
+      cli_args = 'Mesh/gen/background_block_ids="30" Mesh/gen/background_block_names="back"
+                  Mesh/gen/ring_radii="2 3" Mesh/gen/ring_intervals="2 2"
+                  Mesh/gen/ring_inner_boundary_layer_widths="0.0 0.0"
```
why do you allow a width of 0?
seems like this could be forbidden with a rangeCheckedParam

miaoyinb replied:
Allowing width of 0 makes it possible to add boundary layer to some blocks and leave the other blocks free of boundary layers.

GiudGiud commented on modules/reactor/include/meshgenerators/PeripheralRingMeshGenerator.h:
```diff
@@ -30,6 +30,20 @@ class PeripheralRingMeshGenerator : public PolygonMeshGeneratorBase
   const unsigned int _peripheral_layer_num;
+  /// Bias value used to induce biasing to radial meshing in peripheral ring region
+  const Real _peripheral_radial_bias;
+  /// Width of the peripheral ring region to be used as inner boundary layer
+  const Real _peripheral_inner_boundary_layer_width;
+  /// Number of radial sectors of the inner boundary layer mesh of the peripheral ring region
+  const unsigned int _peripheral_inner_boundary_layer_intervals;
+  /// Growth factor used to bias radial meshing of the inner boundary layer of the peripheral ring region
+  const Real _peripheral_inner_boundary_layer_bias;
```
why not use a singleBdryLayerParams?

GiudGiud commented on modules/reactor/include/meshgenerators/PeripheralRingMeshGenerator.h:
```diff
@@ -30,6 +30,20 @@ class PeripheralRingMeshGenerator : public PolygonMeshGeneratorBase
+  const unsigned int _peripheral_inner_boundary_layer_intervals;
+  /// Growth factor used to bias radial meshing of the inner boundary layer of the peripheral ring region
+  const Real _peripheral_inner_boundary_layer_bias;
+  /// Width of the peripheral ring region to be used as outer boundary layer
+  const Real _peripheral_outer_boundary_layer_width;
+  /// Number of radial sectors of the outer boundary layer mesh of the peripheral ring region
+  const unsigned int _peripheral_outer_boundary_layer_intervals;
+  /// Growth factor used to bias radial meshing of the outer boundary layer of the peripheral ring region
+  const Real _peripheral_outer_boundary_layer_bias;
```
same

GiudGiud commented on modules/reactor/include/meshgenerators/PeripheralRingMeshGenerator.h:
```diff
@@ -82,4 +96,28 @@ class PeripheralRingMeshGenerator : public PolygonMeshGeneratorBase
+   * @param input_ext_node_num number of nodes on the external boundary of the input mesh
+   * @param input_bdry_angles list of angles (in rad) formed by three neighboring nodes on the
+   * external boundary of the input mesh
+   * @param ref_inner_bdry_surf Reference outmost layer (surface) points of the inner boundary layer
```
so from  ref_inner_bdry_surf I can get input_bdry_angles and if I know origin_pt I can also get ref_inner_bdry_azi azimuthal ?

miaoyinb replied:
Yes, I may eliminate one

GiudGiud commented on modules/reactor/include/meshgenerators/PeripheralRingMeshGenerator.h:
```diff
@@ -82,4 +96,28 @@ class PeripheralRingMeshGenerator : public PolygonMeshGeneratorBase
+   * boundary layer
+   * @param inner_peripheral_bias_terms terms describing the cumulative radial fractions of the
+   * nodes within the inner boundary layer
+   * @param azi_array list of azimuthal angles (in degrees) for radius correction purpose
```
azimuthal angles of what? How is this different from ref_inner_bdry_azi ?

miaoyinb replied:
azi_array is the list of azi angle of of thenodes on input mesh's external boundary.

By moving these nodes outward along the normal direction, we can get `ref_inner_bdry_azi`. So they are different.

miaoyinb replied:
Will add more description.

GiudGiud commented on modules/reactor/include/meshgenerators/PolygonMeshGeneratorBase.h:
```diff
@@ -65,18 +65,49 @@ class PolygonMeshGeneratorBase : public MeshGenerator
     HEXAGON_NUM_SIDES = 6
   };
 
+  /// A struct type used to contain multiple blocks's boundary layer related parameters
```
```suggestion
  /// Contains multiple blocks's boundary layer related parameters
```

GiudGiud replied:
is block the most descriptive word for them?
It's always a hexagonal ring region right?

GiudGiud commented on modules/reactor/include/meshgenerators/PolygonMeshGeneratorBase.h:
```diff
@@ -65,18 +65,49 @@ class PolygonMeshGeneratorBase : public MeshGenerator
+    std::vector<unsigned int> intervals;
+    std::vector<Real> biases;
+  };
+  /// A struct type used to contain a single block's boundary layer related parameters
```
```suggestion
  /// Contains a single block's boundary layer related parameters
```

GiudGiud commented on modules/reactor/include/meshgenerators/PolygonMeshGeneratorBase.h:
```diff
@@ -65,18 +65,49 @@ class PolygonMeshGeneratorBase : public MeshGenerator
    * Creates a mesh of a slice that corresponds to a single side of the polygon to be generated.
    * @param mesh input mesh to build the slice mesh onto
    * @param ring_radii radii of the ring regions
-   * @param rings numbers of radial intervals of the ring regions
+   * @param ring_layers numbers of radial intervals of the ring regions
+   * @param ring_radial_biases values used for radial meshing biasing in ring regions
```
does the BdryLayerParam not work for regular ring regions?

miaoyinb replied:
regular regions does not need width or fraction attribute.

GiudGiud commented on modules/reactor/include/meshgenerators/PolygonMeshGeneratorBase.h:
```diff
@@ -65,18 +65,49 @@ class PolygonMeshGeneratorBase : public MeshGenerator
+   * factors of the inner boundary layer of the ring regions
+   * @param ring_outer_boundary_layer_params widths, radial fractions, radial sectors, and growth
+   * factors of the outer boundary layer of the ring regions
    * @param ducts_center_dist distance parameters of the duct regions
    * @param ducts_layers numbers of radial intervals of the duct regions
+   * @param duct_radial_biases values used for radial meshing biasing in duct regions
```
same question for ducts

GiudGiud commented on modules/reactor/include/meshgenerators/PolygonMeshGeneratorBase.h:
```diff
@@ -65,18 +65,49 @@ class PolygonMeshGeneratorBase : public MeshGenerator
    * @param has_rings whether the slice contains ring regions or not
    * @param has_ducts whether the slice contains duct regions or not
    * @param num_sectors_per_side number of azimuthal intervals
    * @param background_intervals number of radial intervals of the background region
+   * @param background_radial_bias value used for radial meshing biasing in background region
```
yeah so for background the widths need not be stored explicitly. If you do port rings and ducts to using the struct, let s port background too please!

miaoyinb replied:
I think it is better that regular region does not use struct as neither width nor fraction can be defined on regular background region

GiudGiud commented on modules/reactor/include/meshgenerators/PolygonMeshGeneratorBase.h:
```diff
@@ -401,4 +448,61 @@ class PolygonMeshGeneratorBase : public MeshGenerator
+  std::vector<std::vector<Real>>
+  biasTermsCalculator(const std::vector<Real> radial_biases,
+                      const std::vector<unsigned int> intervals,
+                      const std::vector<Real> inner_boundary_layer_fractions,
+                      const std::vector<unsigned int> inner_boundary_layer_intervals,
+                      const std::vector<Real> inner_boundary_layer_biases,
```
why not pass the struct?

GiudGiud commented on modules/reactor/include/meshgenerators/PolygonMeshGeneratorBase.h:
```diff
@@ -401,4 +448,61 @@ class PolygonMeshGeneratorBase : public MeshGenerator
+   */
+  std::vector<Real> biasTermsCalculator(const Real radial_bias,
+                                        const unsigned int intervals,
+                                        const Real inner_boundary_layer_fraction = 0.0,
+                                        const unsigned int inner_boundary_layer_intervals = 0,
+                                        const Real inner_boundary_layer_bias = 1.0,
```
I guess if you want to use the struct here you d need to add a fraction attribute?

miaoyinb replied:
Yes

GiudGiud reviewed (changes requested):
more comments. I have not looked at the source again, just the headers and tests
Thanks for addressing the previous review!

miaoyinb committed:
use struct for biasTermsCalculator() ref #20637

miaoyinb committed:
eliminate an unnecessary variable ref #20637

clang-format patch 3 ref #20637

Apply suggestions from code review

Minor comments/syntax improvement ref #20637

Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

clang-format fix 4 ref #20637

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1008499/) on f3207a0 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/20638/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/20638/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 4b35627c83e7f9fac90135d0b405f358599c849b`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1008499/) on f3207a0 wanted to post the following:

The following file(s) are changed:

  conda/mpich/conda_build_config.yaml

The conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1008499/) on f3207a0 wanted to post the following:

The following file(s) are changed:

  conda/petsc/meta.yaml

The PETSc conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1008499/) on f3207a0 wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1008546/) on 34e7778 wanted to post the following:

The following file(s) are changed:

  conda/mpich/conda_build_config.yaml

The conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1008546/) on 34e7778 wanted to post the following:

The following file(s) are changed:

  conda/petsc/meta.yaml

The PETSc conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1008546/) on 34e7778 wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [End user mac](https://civet.inl.gov/job/1008565/) on 34e7778 : invalidated by @miaoyinb

GiudGiud commented on modules/reactor/include/meshgenerators/PolygonMeshGeneratorBase.h:
```diff
@@ -401,4 +449,38 @@ class PolygonMeshGeneratorBase : public MeshGenerator
+   * Creates bias terms for multiple blocks.
+   * @param radial_biases bias growth factors of the elements within the main regions of the blocks
+   * @param intervals radial interval numbers of the main regions of the blocks
+   * @ param inner_boundary_layer_params widths, radial fractions, radial sectors, and growth
+   * factors of the inner boundary layers
+   * @ param outer_boundary_layer_params widths, radial fractions, radial sectors, and growth
+   * factors of the outer boundary layers
```
```suggestion
   * @param inner_boundary_layer_params widths, radial fractions, radial sectors, and growth
   * factors of the inner boundary layers
   * @param outer_boundary_layer_params widths, radial fractions, radial sectors, and growth
   * factors of the outer boundary layers
```

GiudGiud commented on modules/reactor/include/meshgenerators/PolygonMeshGeneratorBase.h:
```diff
@@ -401,4 +449,38 @@ class PolygonMeshGeneratorBase : public MeshGenerator
+   * Creates bias terms for a single block.
+   * @param radial_bias bias growth factor of the elements within the main region of the block
+   * @param intervals radial interval number of the main region of the block
+   * @ param inner_boundary_layer_params width, radial fraction, radial sector, and growth
+   * factor of the inner boundary layer
+   * @ param outer_boundary_layer_params width, radial fraction, radial sector, and growth
+   * factor of the outer boundary layer
```
```suggestion
   * @param inner_boundary_layer_params width, radial fraction, radial sector, and growth
   * factor of the inner boundary layer
   * @param outer_boundary_layer_params width, radial fraction, radial sector, and growth
   * factor of the outer boundary layer
```

GiudGiud commented on modules/reactor/src/meshgenerators/PolygonConcentricCircleMeshGeneratorBase.C:
```diff
@@ -25,7 +26,43 @@ PolygonConcentricCircleMeshGeneratorBase::validParams()
       1,
       "background_intervals>0",
-      "Number of radial meshing intervals in background region (area between rings and ducts).");
+      "Number of radial meshing intervals in background region excluding boundary layers (area "
```
```suggestion
      "Number of radial meshing intervals in background region (area "
```

GiudGiud commented on modules/reactor/src/meshgenerators/PolygonConcentricCircleMeshGeneratorBase.C:
```diff
@@ -40,7 +77,35 @@ PolygonConcentricCircleMeshGeneratorBase::validParams()
-      "duct_intervals", "duct_intervals>0", "Number of meshing intervals in each enclosing duct.");
+      "duct_intervals",
+      "duct_intervals>0",
+      "Number of meshing intervals in each enclosing duct excluding boundary layers.");
```
```suggestion
      "Number of meshing intervals in each enclosing duct excluding duct boundary layers.");
```

GiudGiud commented on modules/reactor/src/meshgenerators/PolygonConcentricCircleMeshGeneratorBase.C:
```diff
@@ -25,7 +26,43 @@ PolygonConcentricCircleMeshGeneratorBase::validParams()
       "background_intervals>0",
-      "Number of radial meshing intervals in background region (area between rings and ducts).");
+      "Number of radial meshing intervals in background region excluding boundary layers (area "
+      "between rings and ducts) excluding boundary layers.");
```
```suggestion
      "between rings and ducts) excluding the background's boundary layers.");
```

GiudGiud commented on modules/reactor/src/meshgenerators/PolygonConcentricCircleMeshGeneratorBase.C:
```diff
@@ -50,7 +115,34 @@ PolygonConcentricCircleMeshGeneratorBase::validParams()
       "ring_intervals",
       "ring_intervals>0",
-      "Number of radial mesh intervals within each major concentric circle.");
+      "Number of radial mesh intervals within each major concentric circle excluding boundary "
+      "layers.");
```
```suggestion
      "Number of radial mesh intervals within each major concentric circle excluding their boundary "
      "layers.");
```

GiudGiud commented on modules/reactor/src/meshgenerators/PolygonConcentricCircleMeshGeneratorBase.C:
```diff
@@ -50,7 +115,34 @@ PolygonConcentricCircleMeshGeneratorBase::validParams()
+  params.addRangeCheckedParam<std::vector<Real>>(
+      "ring_inner_boundary_layer_widths",
+      "ring_inner_boundary_layer_widths>=0",
+      "Widths of ring regions that are assigned to be the inner boundary layers.");
```
```suggestion
      "Widths of each ring regions that are assigned to be each ring's inner boundary layers.");
```

GiudGiud commented on modules/reactor/src/meshgenerators/PolygonConcentricCircleMeshGeneratorBase.C:
```diff
@@ -50,7 +115,34 @@ PolygonConcentricCircleMeshGeneratorBase::validParams()
+      "Widths of ring regions that are assigned to be the inner boundary layers.");
+  params.addParam<std::vector<unsigned int>>(
+      "ring_inner_boundary_layer_intervals",
+      "Number of radial intervals of the ring inner boundary layers");
```
```suggestion
      "Number of radial intervals of the rings' inner boundary layers");
```

GiudGiud commented on modules/reactor/src/meshgenerators/PolygonConcentricCircleMeshGeneratorBase.C:
```diff
@@ -50,7 +115,34 @@ PolygonConcentricCircleMeshGeneratorBase::validParams()
+  params.addRangeCheckedParam<std::vector<Real>>(
+      "ring_inner_boundary_layer_biases",
+      "ring_inner_boundary_layer_biases>0",
+      "Growth factors used for mesh biasing of the ring inner boundary layers.");
```
```suggestion
      "Growth factors used for mesh biasing of the rings' inner boundary layers.");
```

GiudGiud commented on modules/reactor/src/meshgenerators/PolygonConcentricCircleMeshGeneratorBase.C:
```diff
@@ -50,7 +115,34 @@ PolygonConcentricCircleMeshGeneratorBase::validParams()
+  params.addRangeCheckedParam<std::vector<Real>>(
+      "ring_outer_boundary_layer_widths",
+      "ring_outer_boundary_layer_widths>=0",
+      "Widths of ring regions that are assigned to be the outer boundary layers.");
```
```suggestion
      "Widths of each ring regions that are assigned to be each ring's outer boundary layers.");
```

GiudGiud commented on modules/reactor/src/meshgenerators/PolygonConcentricCircleMeshGeneratorBase.C:
```diff
@@ -50,7 +115,34 @@ PolygonConcentricCircleMeshGeneratorBase::validParams()
+      "Widths of ring regions that are assigned to be the outer boundary layers.");
+  params.addParam<std::vector<unsigned int>>(
+      "ring_outer_boundary_layer_intervals",
+      "Number of radial intervals of the ring outer boundary layers");
```
```suggestion
      "Number of radial intervals of the rings' outer boundary layers");
```

GiudGiud commented on modules/reactor/src/meshgenerators/PolygonConcentricCircleMeshGeneratorBase.C:
```diff
@@ -50,7 +115,34 @@ PolygonConcentricCircleMeshGeneratorBase::validParams()
+  params.addRangeCheckedParam<std::vector<Real>>(
+      "ring_outer_boundary_layer_biases",
+      "ring_outer_boundary_layer_biases>0",
+      "Growth factors used for mesh biasing of the ring outer boundary layers.");
```
```suggestion
      "Growth factors used for mesh biasing of the rings' outer boundary layers.");
```

GiudGiud commented on modules/reactor/src/meshgenerators/PolygonConcentricCircleMeshGeneratorBase.C:
```diff
@@ -89,7 +181,19 @@ PolygonConcentricCircleMeshGeneratorBase::validParams()
+      "duct_inner_boundary_layer_biases duct_inner_boundary_layer_widths "
+      "duct_inner_boundary_layer_intervals duct_outer_boundary_layer_biases "
+      "duct_outer_boundary_layer_widths duct_outer_boundary_layer_intervals",
+      "Mesh Biasing Options");
```
```suggestion
      "Mesh Boundary Layers and Biasing Options");
```

GiudGiud commented on modules/reactor/src/meshgenerators/PeripheralRingMeshGenerator.C:
```diff
@@ -44,6 +81,12 @@ PeripheralRingMeshGenerator::validParams()
+      "peripheral_inner_boundary_layer_intervals peripheral_inner_boundary_layer_bias "
+      "peripheral_outer_boundary_layer_width peripheral_outer_boundary_layer_intervals "
+      "peripheral_outer_boundary_layer_bias",
+      "Mesh Biasing Options");
```
```suggestion
      "Mesh Boundary Layer and Biasing Options");
```

GiudGiud reviewed (commented):
looks ready to me after this

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1016507/) on f892cb0 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/20638/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/20638/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format c5afa6f300ae21fb1847eaeb94e26df24a48b02d`

GiudGiud commented:
@miaoyinb could you please fix-up the commit history?
Let's squash all the "suggestions"/"typo" etc small commits.

miaoyinb commented:
> @miaoyinb could you please fix-up the commit history? Let's squash all the "suggestions"/"typo" etc small commits.

You want to squash things into a single commit?

miaoyinb commented:
> > @miaoyinb could you please fix-up the commit history? Let's squash all the "suggestions"/"typo" etc small commits.
> 
> You want to squash things into a single commit?

I squashed 10 "minor" commits. Let me know if further squashing is needed. Thanks.

moosebuild commented:
Job [Test install](https://civet.inl.gov/job/1016879/) on 7111764 : invalidated by @miaoyinb

moosebuild commented:
Job [Test install](https://civet.inl.gov/job/1016879/) on 7111764 : invalidated by @miaoyinb

GiudGiud commented:
Thanks for the PR!

miaoyinb commented:
Thank you so much, @GiudGiud
